### ISSUE TITLE
Using createAction, how would one go about validating the payload passed in?

### ISSUE THREAD

**Issue by rmisio** [reduxjs/redux-starter-kit#146]
Let's say I'd like to ensure some required properties are passed into the action creator. How would I do that?

I considered a wrapper function, but then the toString is not on the wrapper and would have to be carried over.

```
export const activateConvo = peerId => {
  if (typeof peerId !== 'string' || !peerId) {
    throw new Error('Please provide a peerId as a non-empty string.');
  }

  return createAction('CHAT_CONVO_ACTIVATE')(peerId);
}
```

Forgive me if this is more of a StackOverflow type question... but not really seeing any redux-starter-kit talk there.

**Comment by markerikson** [comment]
That is kind of the tradeoff atm, yeah.  If you are using `createAction()` (whether directly or via `createSlice()`), then you are basically saying you're fine with its standard way of doing things.

You _can_ always still write action creators by hand if you want to.  The point of the functions from RSK is to provide an opinionated, obvious, no-boilerplate way of doing things.

That said, I'm still considering the idea of a "payload customization" callback somehow - I just haven't had time to think about how that might fit in.

**Comment by CGamesPlay** [comment]
I'd like to vote for this issue as well. My proposal would literally be a payload transformer:

```
createAction('CHAT_CONVO_ACTIVATE', (peerId) => {
  if (typeof peerId !== 'string' || !peerId) {
    throw new Error('Please provider a peerId as a non-empty string.');
  }
  return peerId;
});
```

In this case, the transformer is the identity function with some validation. My use case would be to change it into a different value.

**Comment by markerikson** [comment]
@CGamesPlay : yeah, that's basically the kind of thing I've been picturing.  The biggest issue is how to make use of that with `createReducer` and `createSlice`.

**Comment by CGamesPlay** [comment]
I'm using this in my project that doesn't use `createReducer` (note: I haven't thought about the `mapper === undefined` case, the types are likely a bit weird there).

```
export interface ActionCreator<T extends string, P, A extends any[]> {
  (...args: A): PayloadAction<P, T>;
  type: T;
  matches(a: Action<T>): a is PayloadAction<P, T>;
}

export const createAction = <T extends string, A extends any[], P>(
  type: T,
  payload: (...args: A) => P = undefined,
): ActionCreator<T, P, A> => {
  const creator: any = payload
    ? (...args: A) => ({ type, payload: payload(...args) })
    : () => ({ type });
  creator.toString = () => name;
  creator.type = type;
  creator.matches = a => a && a.type === type;
  return creator;
};

// Typed action payloads in my custom reducers:

function customReducer(state, action) {
  if (activateConvo.matches(action)) {
    // action.payload is number
  }
}
```

Is the concern that TypeScript won't automatically know the action type in your `createReducer` reducer? Getting TypeScript to automatically know that might not be possible with the current API, but perhaps you can create a utility type that extracts the correct type? Something like this:

```
createReducer(0, {
  [increment.type]: (state, action: ActionType<increment>) => state + action.payload,
});
```

**Comment by markerikson** [comment]
Fixed by #149 and out as 0.6.3 .  Please let me know if that works for you!
